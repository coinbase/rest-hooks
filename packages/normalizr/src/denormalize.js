import * as ImmutableUtils from './schemas/ImmutableUtils';
import * as ArrayUtils from './schemas/Array';
import * as ObjectUtils from './schemas/Object';

const unvisitEntity = (id, schema, unvisit, getEntity, cache) => {
  const entity = getEntity(id, schema);
  if (typeof entity !== 'object' || entity === null) {
    return [entity, false];
  }

  if (!cache[schema.key]) {
    cache[schema.key] = {};
  }

  let found = true;
  if (!cache[schema.key][id]) {
    // Ensure we don't mutate it non-immutable objects
    const entityCopy = ImmutableUtils.isImmutable(entity)
      ? entity
      : { ...entity };

    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    cache[schema.key][id] = entityCopy;
    [cache[schema.key][id], found] = schema.denormalize(entityCopy, unvisit);
  }

  return [cache[schema.key][id], found];
};

const getUnvisit = entities => {
  const cache = {};
  const getEntity = getEntities(entities);

  return [
    function unvisit(input, schema) {
      if (!schema) return [input, true];

      if (
        typeof schema === 'object' &&
        (!schema.denormalize || typeof schema.denormalize !== 'function')
      ) {
        const method = Array.isArray(schema)
          ? ArrayUtils.denormalize
          : ObjectUtils.denormalize;
        return method(schema, input, unvisit);
      }

      // null is considered intentional, thus always 'found' as true
      if (input === null) {
        return [input, true];
      }

      if (
        typeof schema.getId === 'function' &&
        typeof schema.normalize === 'function'
      ) {
        // unvisitEntity just can't handle undefined
        if (input === undefined) {
          return [input, false];
        }
        return unvisitEntity(input, schema, unvisit, getEntity, cache);
      }

      if (typeof schema.denormalize === 'function') {
        return schema.denormalize(input, unvisit);
      }

      return [input, true];
    },
    cache,
  ];
};

const getEntities = entities => {
  const isImmutable = ImmutableUtils.isImmutable(entities);

  return (entityOrId, schema) => {
    const schemaKey = schema.key;

    if (typeof entityOrId === 'object') {
      return entityOrId;
    }

    if (isImmutable) {
      return entities.getIn([schemaKey, entityOrId.toString()]);
    }

    return entities[schemaKey] && entities[schemaKey][entityOrId];
  };
};

export const denormalize = (input, schema, entities) => {
  /* istanbul ignore next */
  // eslint-disable-next-line no-undef
  if (process.env.NODE_ENV !== 'production' && schema === undefined)
    throw new Error('shema needed');
  if (typeof input !== 'undefined') {
    const [unvisit, cache] = getUnvisit(entities);
    return [...unvisit(input, schema), cache];
  }
  return [undefined, false, {}];
};

export const denormalizeSimple = (input, schema, entities) =>
  denormalize(input, schema, entities).slice(0, 2);
