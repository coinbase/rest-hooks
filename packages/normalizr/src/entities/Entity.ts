import SimpleRecord, { SimpleRecordSchema } from './SimpleRecord';
import { isImmutable, denormalizeImmutable } from '../schemas/ImmutableUtils';
import * as schema from '../schema';
import { AbstractInstanceType, Schema } from '../types';

/** Represents data that should be deduped by specifying a primary key. */
export default abstract class Entity extends SimpleRecord {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Returns the globally unique identifier for the static Entity */
  static get key(): string {
    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'production' && this.name === '')
      throw new Error(
        'Entity classes without a name must define `static get key()`',
      );
    return this.name;
  }

  /** Defines indexes to enable lookup by */
  declare static indexes?: readonly string[];

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  static pk<T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ): string | undefined {
    return this.prototype.pk.call(value, parent, key) || key;
  }

  static normalize(
    input: any,
    parent: any,
    key: string | undefined,
    visit: Function,
    addEntity: Function,
    visitedEntities: Record<string, any>,
  ): any {
    // TODO: what's store needs to be a differing type from fromJS
    const processedEntity = this.fromJS(input, parent, key);
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      const instanceSample = new (this as any)();
      const keysOfRecord = new Set(Object.keys(instanceSample));
      let [sameCount, diffCount] = [0, 0];
      let extraKey = false;
      for (const keyOfProps of this.keysDefined(processedEntity)) {
        if (keysOfRecord.has(keyOfProps)) {
          sameCount++;
        } else {
          diffCount++;
          extraKey = true;
        }
      }
      diffCount += keysOfRecord.size - sameCount;
      if (diffCount > sameCount && keysOfRecord.size && extraKey) {
        const error = new Error(
          `Attempted to initialize ${
            this.name
          } with substantially different than expected keys

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Expected keys: ${Object.keys(instanceSample)}
  Value: ${JSON.stringify(this.toObjectDefined(processedEntity), null, 2)}`,
        );
        (error as any).status = 400;
        throw error;
      }
    }
    const id = processedEntity.pk(parent, key);
    if (id === undefined || id === '') {
      /* istanbul ignore else */
      if (process.env.NODE_ENV !== 'production') {
        const error = new Error(
          `Missing usable resource key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Entity: ${this.name}
  Value: ${input && JSON.stringify(input, null, 2)}
  `,
        );
        (error as any).status = 400;
        throw error;
      } else {
        throw new Error('undefined pk');
      }
    }
    const entityType = this.key;

    if (!(entityType in visitedEntities)) {
      visitedEntities[entityType] = {};
    }
    if (!(id in visitedEntities[entityType])) {
      visitedEntities[entityType][id] = [];
    }
    if (
      visitedEntities[entityType][id].some((entity: any) => entity === input)
    ) {
      return id;
    }
    visitedEntities[entityType][id].push(input);

    Object.keys(this.schema).forEach(key => {
      if (
        Object.hasOwnProperty.call(processedEntity, key) &&
        typeof processedEntity[key] === 'object'
      ) {
        const schema = this.schema[key];
        processedEntity[key] = visit(
          processedEntity[key],
          processedEntity,
          key,
          schema,
          addEntity,
          visitedEntities,
        );
      }
    });

    addEntity(this, processedEntity, processedEntity, parent, key);
    return id;
  }

  static denormalize<T extends typeof SimpleRecord>(
    this: T,
    entity: AbstractInstanceType<T>,
    unvisit: schema.UnvisitFunction,
  ): [AbstractInstanceType<T>, boolean] {
    if (isImmutable(entity)) {
      const [denormEntity, found] = denormalizeImmutable(
        this.schema,
        entity,
        unvisit,
      );
      return [this.fromJS(denormEntity.toObject()), found];
    }
    let found = true;
    const denormEntity = entity;

    Object.keys(this.schema).forEach(key => {
      const schema = this.schema[key];
      const [value, foundItem] = unvisit(entity[key], schema);
      if (!foundItem) {
        found = false;
      }
      if (this.hasDefined(entity, key as any) && denormEntity[key] !== value) {
        denormEntity[key] = value;
      }
    });

    return [denormEntity as any, found];
  }
}

/* istanbul ignore next */
if (process.env.NODE_ENV !== 'production') {
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof SimpleRecord>(
    this: T,
    props: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === undefined)
      throw new Error('cannot construct on abstract types');
    return SimpleRecord.fromJS.call(this, props) as any;
  };
}

export type EntitySchema<E extends typeof SimpleRecord> = SimpleRecordSchema<E>;

export function isEntity(schema: Schema | null): schema is typeof Entity {
  return schema !== null && (schema as any).pk !== undefined;
}
