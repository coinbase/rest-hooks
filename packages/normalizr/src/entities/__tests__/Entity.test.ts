// eslint-env jest
import { fromJS, Record } from 'immutable';
import { first } from 'lodash';

import { denormalizeSimple as denormalize } from '../../denormalize';
import { normalize, schema, SimpleRecord, AbstractInstanceType } from '../../';
import Entity from '../Entity';
import IDEntity from '../IDEntity';

const values = obj => Object.keys(obj).map(key => obj[key]);

class Tacos extends IDEntity {
  readonly name: string = '';
  readonly alias: string | undefined = undefined;
}

describe(`${Entity.name} normalization`, () => {
  test('normalizes an entity', () => {
    class MyEntity extends IDEntity {}
    expect(normalize({ id: '1' }, MyEntity)).toMatchSnapshot();
  });

  it('should throw a custom error if data does not include pk', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      pk() {
        return this.name;
      }
    }
    const schema = MyEntity;
    function normalizeBad() {
      normalize({ secondthing: 'hi' }, schema);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  it('should throw a custom error if data loads with no matching props', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      pk() {
        return this.name;
      }
    }
    const schema = MyEntity;
    function normalizeBad() {
      normalize({}, schema);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  it('should not expect getters returned', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      get other() {
        return this.name + 5;
      }

      get another() {
        return 'another';
      }

      get yetAnother() {
        return 'another2';
      }

      pk() {
        return this.name;
      }
    }
    function normalizeBad() {
      normalize({ name: 'bob' }, MyEntity);
    }
    expect(normalizeBad).not.toThrow();
  });

  it('should throw if data loads with unexpected prop that is a getter', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      readonly thirdthing: number = 0;

      get nonexistantthing() {
        return this.name + 5;
      }

      pk() {
        return this.name;
      }
    }
    function normalizeBad() {
      normalize({ name: 'hoho', nonexistantthing: 'hi' }, MyEntity);
    }
    expect(normalizeBad).toThrow();
  });

  it('should throw if data loads with unexpected prop that is a method', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      readonly thirdthing: number = 0;
      readonly thirdthing2: number = 0;

      nonexistantthing() {
        return this.name + 5;
      }

      nonexistantthing2() {
        return this.name + 5;
      }

      nonexistantthing3() {
        return this.name + 5;
      }

      nonexistantthing4() {
        return this.name + 5;
      }

      pk() {
        return this.name;
      }
    }
    function normalizeBad() {
      normalize(
        {
          name: 'hoho',
          nonexistantthing: 'hi',
          nonexistantthing2: 'hi',
          nonexistantthing3: 'hi',
        },
        MyEntity,
      );
    }
    expect(normalizeBad).toThrow();
  });

  it('should throw a custom error if data loads with half unexpected props', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      readonly thirdthing: number = 0;
      pk() {
        return this.name;
      }
    }
    const schema = MyEntity;
    function normalizeBad() {
      normalize({ name: 'hoho', nonexistantthing: 'hi' }, schema);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  it('should throw a custom error if data loads with string', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      readonly thirdthing: number = 0;
      pk() {
        return this.name;
      }
    }
    const schema = MyEntity;
    function normalizeBad() {
      normalize('hibho', schema);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  describe('key', () => {
    test('must be created with a key name', () => {
      const makeSchema = () =>
        class extends Entity {
          readonly id: number = 0;
          pk() {
            return `${this.id}`;
          }
        };
      expect(() => makeSchema().key).toThrow();
    });

    test('key name must be a string', () => {
      // @ts-expect-error
      class MyEntity extends IDEntity {
        static get key() {
          return 42;
        }
      }
    });

    test('key getter should return key set via `static get key()`', () => {
      class User extends IDEntity {
        static get key() {
          return 'user';
        }
      }
      expect(User.key).toEqual('user');
    });
  });

  describe('pk()', () => {
    test('can use a custom pk() string', () => {
      class User extends Entity {
        readonly idStr: string = '';
        readonly name: string = '';

        pk() {
          return this.idStr;
        }
      }
      expect(
        normalize({ idStr: '134351', name: 'Kathy' }, User),
      ).toMatchSnapshot();
    });

    test('can normalize entity IDs based on their object key', () => {
      class User extends Entity {
        readonly name: string = '';
        pk(parent?: any, key?: string) {
          return key;
        }
      }
      const inputSchema = new schema.Values({ users: User }, () => 'users');

      expect(
        normalize(
          { '4': { name: 'taco' }, '56': { name: 'burrito' } },
          inputSchema,
        ),
      ).toMatchSnapshot();
    });

    test("can build the entity's ID from the parent object", () => {
      class User extends Entity {
        readonly id: string = '';
        readonly name: string = '';
        pk(parent?: any, key?: string) {
          return `${parent.name}-${key}-${this.id}`;
        }
      }
      const inputSchema = new schema.Object({ user: User });

      expect(
        normalize(
          { name: 'tacos', user: { id: '4', name: 'Jimmy' } },
          inputSchema,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('mergeStrategy', () => {
    test('defaults to plain merging', () => {
      expect(
        normalize(
          [
            { id: '1', name: 'foo' },
            { id: '1', name: 'bar', alias: 'bar' },
          ],
          [Tacos],
        ),
      ).toMatchSnapshot();
    });

    test('can use a custom merging strategy', () => {
      class MergeTaco extends Tacos {
        static merge<T extends typeof SimpleRecord>(
          this: T,
          existing: AbstractInstanceType<T>,
          incoming: AbstractInstanceType<T>,
        ) {
          const props = Object.assign(
            {},
            this.toObjectDefined(existing),
            this.toObjectDefined(incoming),
            { name: (existing as MergeTaco).name },
          );
          return this.fromJS(props);
        }
      }

      expect(
        normalize(
          [
            { id: '1', name: 'foo' },
            { id: '1', name: 'bar', alias: 'bar' },
          ],
          [MergeTaco],
        ),
      ).toMatchSnapshot();
    });
  });

  describe('processStrategy', () => {
    test('can use a custom processing strategy', () => {
      class ProcessTaco extends Tacos {
        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...props,
            slug: `thing-${(props as ProcessTaco).id}`,
          }) as AbstractInstanceType<T>;
        }
      }

      expect(
        normalize({ id: '1', name: 'foo' }, ProcessTaco),
      ).toMatchSnapshot();
    });

    test('can use information from the parent in the process strategy', () => {
      class ChildEntity extends IDEntity {
        readonly content: string = '';
        readonly parentId: string = '';
        readonly parentKey: string = '';

        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...props,
            parentId: parent.id,
            parentKey: key,
          }) as AbstractInstanceType<T>;
        }
      }
      class ParentEntity extends IDEntity {
        readonly content: string = '';

        static schema = { child: ChildEntity };
      }

      expect(
        normalize(
          {
            id: '1',
            content: 'parent',
            child: { id: '4', content: 'child' },
          },
          ParentEntity,
        ),
      ).toMatchSnapshot();
    });

    test('is run before and passed to the schema normalization', () => {
      class AttachmentsEntity extends IDEntity {}
      class EntriesEntity extends IDEntity {
        readonly type: string = '';

        static schema = {
          data: { attachment: AttachmentsEntity },
        };

        static fromJS<T extends typeof SimpleRecord>(
          this: T,
          props: Partial<AbstractInstanceType<T>>,
          parent?: any,
          key?: string,
        ) {
          return super.fromJS({
            ...values(props)[0],
            type: Object.keys(props)[0],
          }) as AbstractInstanceType<T>;
        }
      }

      expect(
        normalize(
          { message: { id: '123', data: { attachment: { id: '456' } } } },
          EntriesEntity,
        ),
      ).toMatchSnapshot();
    });
  });
});

describe(`${Entity.name} denormalization`, () => {
  test('denormalizes an entity', () => {
    const entities = {
      Tacos: {
        '1': { id: '1', type: 'foo' },
      },
    };
    expect(denormalize('1', Tacos, entities)).toMatchSnapshot();
    expect(denormalize('1', Tacos, fromJS(entities))).toMatchSnapshot();
  });

  class Food extends IDEntity {}
  class Menu extends IDEntity {
    readonly food: Food = Food.fromJS();

    static schema = { food: Food };
  }

  test('denormalizes deep entities', () => {
    const entities = {
      Menu: {
        '1': { id: '1', food: '1' },
        '2': { id: '2' },
      },
      Food: {
        '1': { id: '1' },
      },
    };

    expect(denormalize('1', Menu, entities)).toMatchSnapshot();
    expect(denormalize('1', Menu, fromJS(entities))).toMatchSnapshot();

    expect(denormalize('2', Menu, entities)).toMatchSnapshot();
    expect(denormalize('2', Menu, fromJS(entities))).toMatchSnapshot();
  });

  test('denormalizes to undefined for missing data', () => {
    const entities = {
      Menu: {
        '1': { id: '1', food: '2' },
      },
      Food: {
        '1': { id: '1' },
      },
    };

    expect(denormalize('1', Menu, entities)).toMatchSnapshot();
    expect(denormalize('1', Menu, fromJS(entities))).toMatchSnapshot();

    expect(denormalize('2', Menu, entities)).toMatchSnapshot();
    expect(denormalize('2', Menu, fromJS(entities))).toMatchSnapshot();
  });

  test('denormalizes deep entities with records', () => {
    const Food = Record({ id: null });
    const MenuR = Record({ id: null, food: null });

    const entities = {
      Menu: {
        '1': new MenuR({ id: '1', food: '1' }),
        '2': new MenuR({ id: '2' }),
      },
      Food: {
        '1': new Food({ id: '1' }),
      },
    };

    expect(denormalize('1', Menu, entities)).toMatchSnapshot();
    expect(denormalize('1', Menu, fromJS(entities))).toMatchSnapshot();

    expect(denormalize('2', Menu, entities)).toMatchSnapshot();
    expect(denormalize('2', Menu, fromJS(entities))).toMatchSnapshot();
  });

  test('can denormalize already partially denormalized data', () => {
    const entities = {
      Menu: {
        '1': { id: '1', food: { id: '1' } },
      },
      Food: {
        '1': { id: '1' },
      },
    };

    expect(denormalize('1', Menu, entities)).toMatchSnapshot();
    expect(denormalize('1', Menu, fromJS(entities))).toMatchSnapshot();
  });

  class User extends IDEntity {
    readonly role = '';
    readonly reports: Report[] = [];
  }
  class Report extends IDEntity {
    readonly title: string = '';
    readonly draftedBy: User = User.fromJS();
    readonly publishedBy: User = User.fromJS();

    static schema = {
      draftedBy: User,
      publishedBy: User,
    };
  }
  User.schema = {
    reports: [Report],
  };

  test('denormalizes recursive dependencies', () => {
    const entities = {
      Report: {
        '123': {
          id: '123',
          title: 'Weekly report',
          draftedBy: '456',
          publishedBy: '456',
        },
      },
      User: {
        '456': {
          id: '456',
          role: 'manager',
          reports: ['123'],
        },
      },
    };

    expect(denormalize('123', Report, entities)).toMatchSnapshot();
    expect(denormalize('123', Report, fromJS(entities))).toMatchSnapshot();

    expect(denormalize('456', User, entities)).toMatchSnapshot();
    expect(denormalize('456', User, fromJS(entities))).toMatchSnapshot();
  });

  test('denormalizes entities with referential equality', () => {
    const entities = {
      Report: {
        '123': {
          id: '123',
          title: 'Weekly report',
          draftedBy: '456',
          publishedBy: '456',
        },
      },
      User: {
        '456': {
          id: '456',
          role: 'manager',
          reports: ['123'],
        },
      },
    };

    const [denormalizedReport] = denormalize('123', Report, entities);

    expect(denormalizedReport).toBeDefined();
    // This is just for TypeScript, the above line actually determines this
    if (!denormalizedReport) throw new Error('expected to be defined');
    expect(denormalizedReport).toBe(denormalizedReport.draftedBy?.reports[0]);
    expect(denormalizedReport.publishedBy).toBe(denormalizedReport.draftedBy);

    // NOTE: Given how immutable data works, referential equality can't be
    // maintained with nested denormalization.
  });
});
