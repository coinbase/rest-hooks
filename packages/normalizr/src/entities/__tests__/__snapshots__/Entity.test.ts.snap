// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity denormalization can denormalize already partially denormalized data 1`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization can denormalize already partially denormalized data 2`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes an entity 1`] = `
Array [
  Tacos {
    "alias": undefined,
    "id": "1",
    "name": "",
    "type": "foo",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes an entity 2`] = `
Array [
  Tacos {
    "alias": undefined,
    "id": "1",
    "name": "",
    "type": "foo",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities 1`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities 2`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities 3`] = `
Array [
  Menu {
    "food": Food {
      "id": undefined,
    },
    "id": "2",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities 4`] = `
Array [
  Menu {
    "food": Food {
      "id": undefined,
    },
    "id": "2",
  },
  false,
]
`;

exports[`Entity denormalization denormalizes deep entities with records 1`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities with records 2`] = `
Array [
  Menu {
    "food": Food {
      "id": "1",
    },
    "id": "1",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities with records 3`] = `
Array [
  Menu {
    "food": null,
    "id": "2",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes deep entities with records 4`] = `
Array [
  Menu {
    "food": null,
    "id": "2",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes recursive dependencies 1`] = `
Array [
  Report {
    "draftedBy": User {
      "id": "456",
      "reports": Array [
        [Circular],
      ],
      "role": "manager",
    },
    "id": "123",
    "publishedBy": User {
      "id": "456",
      "reports": Array [
        [Circular],
      ],
      "role": "manager",
    },
    "title": "Weekly report",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes recursive dependencies 2`] = `
Array [
  Report {
    "draftedBy": User {
      "id": "456",
      "reports": Immutable.List [
        Immutable.Map {
          "id": "123",
          "title": "Weekly report",
          "draftedBy": "456",
          "publishedBy": "456",
        },
      ],
      "role": "manager",
    },
    "id": "123",
    "publishedBy": User {
      "id": "456",
      "reports": Immutable.List [
        Immutable.Map {
          "id": "123",
          "title": "Weekly report",
          "draftedBy": "456",
          "publishedBy": "456",
        },
      ],
      "role": "manager",
    },
    "title": "Weekly report",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes recursive dependencies 3`] = `
Array [
  User {
    "id": "456",
    "reports": Array [
      Report {
        "draftedBy": [Circular],
        "id": "123",
        "publishedBy": [Circular],
        "title": "Weekly report",
      },
    ],
    "role": "manager",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes recursive dependencies 4`] = `
Array [
  User {
    "id": "456",
    "reports": Immutable.List [
      Report {
        "draftedBy": Immutable.Map {
          "id": "456",
          "role": "manager",
          "reports": Immutable.List [
            "123",
          ],
        },
        "id": "123",
        "publishedBy": Immutable.Map {
          "id": "456",
          "role": "manager",
          "reports": Immutable.List [
            "123",
          ],
        },
        "title": "Weekly report",
      },
    ],
    "role": "manager",
  },
  true,
]
`;

exports[`Entity denormalization denormalizes to undefined for missing data 1`] = `
Array [
  Menu {
    "food": undefined,
    "id": "1",
  },
  false,
]
`;

exports[`Entity denormalization denormalizes to undefined for missing data 2`] = `
Array [
  Menu {
    "food": undefined,
    "id": "1",
  },
  false,
]
`;

exports[`Entity denormalization denormalizes to undefined for missing data 3`] = `
Array [
  undefined,
  false,
]
`;

exports[`Entity denormalization denormalizes to undefined for missing data 4`] = `
Array [
  undefined,
  false,
]
`;

exports[`Entity normalization mergeStrategy can use a custom merging strategy 1`] = `
Object {
  "entities": Object {
    "MergeTaco": Object {
      "1": MergeTaco {
        "alias": "bar",
        "id": "1",
        "name": "foo",
      },
    },
  },
  "indexes": Object {},
  "result": Array [
    "1",
    "1",
  ],
}
`;

exports[`Entity normalization mergeStrategy defaults to plain merging 1`] = `
Object {
  "entities": Object {
    "Tacos": Object {
      "1": Tacos {
        "alias": "bar",
        "id": "1",
        "name": "bar",
      },
    },
  },
  "indexes": Object {},
  "result": Array [
    "1",
    "1",
  ],
}
`;

exports[`Entity normalization normalizes an entity 1`] = `
Object {
  "entities": Object {
    "MyEntity": Object {
      "1": MyEntity {
        "id": "1",
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization pk() can build the entity's ID from the parent object 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "tacos-user-4": User {
        "id": "4",
        "name": "Jimmy",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "name": "tacos",
    "user": "tacos-user-4",
  },
}
`;

exports[`Entity normalization pk() can normalize entity IDs based on their object key 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "4": User {
        "name": "taco",
      },
      "56": User {
        "name": "burrito",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "4": Object {
      "id": "4",
      "schema": "users",
    },
    "56": Object {
      "id": "56",
      "schema": "users",
    },
  },
}
`;

exports[`Entity normalization pk() can use a custom pk() string 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "134351": User {
        "idStr": "134351",
        "name": "Kathy",
      },
    },
  },
  "indexes": Object {},
  "result": "134351",
}
`;

exports[`Entity normalization processStrategy can use a custom processing strategy 1`] = `
Object {
  "entities": Object {
    "ProcessTaco": Object {
      "1": ProcessTaco {
        "alias": undefined,
        "id": "1",
        "name": "foo",
        "slug": "thing-1",
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization processStrategy can use information from the parent in the process strategy 1`] = `
Object {
  "entities": Object {
    "ChildEntity": Object {
      "4": ChildEntity {
        "content": "child",
        "id": "4",
        "parentId": "1",
        "parentKey": "child",
      },
    },
    "ParentEntity": Object {
      "1": ParentEntity {
        "child": "4",
        "content": "parent",
        "id": "1",
      },
    },
  },
  "indexes": Object {},
  "result": "1",
}
`;

exports[`Entity normalization processStrategy is run before and passed to the schema normalization 1`] = `
Object {
  "entities": Object {
    "AttachmentsEntity": Object {
      "456": AttachmentsEntity {
        "id": "456",
      },
    },
    "EntriesEntity": Object {
      "123": EntriesEntity {
        "data": Object {
          "attachment": "456",
        },
        "id": "123",
        "type": "message",
      },
    },
  },
  "indexes": Object {},
  "result": "123",
}
`;

exports[`Entity normalization should throw a custom error if data does not include pk 1`] = `
"Missing usable resource key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Entity: MyEntity
  Value: {
  \\"secondthing\\": \\"hi\\"
}
  "
`;

exports[`Entity normalization should throw a custom error if data loads with half unexpected props 1`] = `
"Attempted to initialize MyEntity with substantially different than expected keys

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Expected keys:
    Found: name
    Missing: secondthing,thirdthing
  Unexpected keys: nonexistantthing
  Value: {
  \\"name\\": \\"hoho\\",
  \\"nonexistantthing\\": \\"hi\\"
}"
`;

exports[`Entity normalization should throw a custom error if data loads with no matching props 1`] = `
"Attempted to initialize MyEntity with substantially different than expected keys

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.

  Expected keys:
    Found: 
    Missing: name,secondthing
  Unexpected keys: 
  Value: {}"
`;
